# .github/workflows/create-container-functional-tests.yml
name: Functional Tests

on:
  workflow_call:
    inputs:
      type:
        required: true
        type: string
      model:
        required: true
        type: string

jobs:
  functional-test:
    runs-on: self-hosted
    container:
      image: localhost:5000/flagscale_cicd:v1.5
      ports:
        - 80
      volumes:
        - /home/flagscale_cicd/flask/static:/workspace/report
        - /home/flagscale_cicd/docker/docker_build/docker_data:/home/gitlab-runner/data
        - /home/flagscale_cicd/docker/docker_build/docker_tokenizers:/home/gitlab-runner/tokenizers
      options: --gpus all --shm-size=500g --hostname flagscale_cicd
    
    steps:
      - name: Checkout Code
        uses: actions/checkout@v4
        with:
          repository: ${{ github.event.pull_request.head.repo.full_name }} 
          ref: ${{ github.event.pull_request.head.ref }}
          ssh-strict: true
          ssh-user: git
          persist-credentials: true
          clean: true
          sparse-checkout-cone-mode: true
          fetch-tags: false
          show-progress: true
          lfs: false
          submodules: false
          set-safe-directory: true

      - name: Run Functional Test
        run: |
          max_attempts=10
          attempts=0
          success=0

          while [ $attempts -lt $max_attempts ]; do
              echo "Attempt $((attempts + 1)) of $max_attempts"
              
              # Run the functional test script
              tests/scripts/functional_tests/test_model.sh --type ${{ inputs.type }} --model ${{ inputs.model }} || true
              
              if [ $? -eq 0 ]; then
                  echo "Functional test succeeded on attempt $((attempts + 1))"
                  success=1
                  break
              else
                  echo "Functional test failed on attempt $((attempts + 1))"
                  attempts=$((attempts + 1))
              fi
          done

          # If all attempts failed, exit with an error
          if [ $success -eq 0 ]; then
              echo "Error: All $max_attempts attempts failed"
              exit 1
          fi
